### ISSUE TITLE
"The provided execution role does not have permissions to call Publish on SNS."

### ISSUE THREAD

**Issue by theburningmonk** [functionalone/serverless-iam-roles-per-function#5]
Hi ya, ran into another error as I try to integrate this plugin into the demo app I'm building for my course, which has API, Kinesis and SNS functions.

As you can see in [this commit](https://github.com/theburningmonk/manning-aws-lambda-in-motion/commit/6df8557b23051eb44e583ad0230e7758275a65a1), I'm moving everything over to per-function roles, but when I run `sls deploy` I get an error from CloudFormation saying `An error occurred: RetryDashnotifyDashuserLambdaFunction - The provided execution role does not have permissions to callPublish on SNS.`

I have looked at the CF template SLS generates, and the role does include the `sns:Publish` permission.
![image](https://user-images.githubusercontent.com/546969/36634882-7a9df28a-19a3-11e8-96e8-2202e6963b2d.png)

If I do the following then it works:

1. comment out the two SNS functions' `iamRoleStatements` so they use the default `IamRoleLambdaExecution` role that SLS generates, but leave all the other functions with their per-function IAM roles
2. deploy
3. uncomment the two SNS functions' `iamRoleStatements`
4. deploy

Based on the above, it feels like there's some race condition with creating these SNS functions, but I can't quite put my finger on what yet. Thoughts?

**Comment by glicht** [comment]
Thanks for reporting this. It sounds like the issue is with the dlq sns topic. From what I saw in your commit the function is configured with onError. I just noticed that when configuring a sns dlq there is need to add the proper `sns:Publish` permission for the sns dlq. 

Can you try adding to the function `retry-notify-user` the following permission:
```yaml
- Effect: Allow
  Action: sns:Publish
  Resource:
    - arn:aws:sns:#{AWS::Region}:#{AWS::AccountId}:user-notification-dlq
```

I think this should allow the deploy to go through. I'll check about having this permission auto added when a sns dlq is used.

**Comment by theburningmonk** [comment]
That's a good catch, didn't think about the DLQ topic, the CF error should mention the name of the topic in this case (I'll feed that back to them).

And adding the DLQ permission fixes the problem, thanks a lot 👍

**Comment by glicht** [comment]
Excellent. I am looking into a fix for this so the proper permission will be added automatically when using a function level role. Note that I just discovered a related bug in the serverless framework that when a Ref is used to the sns topic for DLQ, the iam role is not updated. See: https://github.com/serverless/serverless/issues/4770

**Comment by glicht** [comment]
Just pushed a fix which should resolve this. Also published version 0.1.5 to npmjs with the fix.

**Comment by theburningmonk** [comment]
Awesome, thanks again for such quick turn-around!
